service/history/workflow/mutable_state_impl.go

```diff
@@ -2815,6 +2815,34 @@ func (e *MutableStateImpl) ReplicateUpsertWorkflowSearchAttributesEvent(
 	e.executionInfo.SearchAttributes = mergeMapOfPayload(currentSearchAttr, upsertSearchAttr)
 }
 
+func (e *MutableStateImpl) AddUpsertWorkflowMemoEvent(
+	workflowTaskCompletedEventID int64,
+	command *commandpb.UpsertWorkflowMemoCommandAttributes,
+) (*historypb.HistoryEvent, error) {
+	opTag := tag.WorkflowActionUpsertWorkflowMemo
+	if err := e.checkMutability(opTag); err != nil {
+		return nil, err
+	}
+
+	event := e.hBuilder.AddUpsertWorkflowMemoEvent(workflowTaskCompletedEventID, command)
+	e.ReplicateUpsertWorkflowMemoEvent(event)
+	// TODO merge active & passive task generation
+	if err := e.taskGenerator.GenerateWorkflowSearchAttrTasks(
```

Comment:
`GenerateWorkflowMemoTasks`?

Reply:
It should be called `GenerateUpsertVisibilityTask`.

Reply:
I just saw that generate memo tasks method is added in task generator but not used here. Not actually trying to comment on the naming. 

But yeah, we should renaming the existing method on task generator instead of adding a new one.

Reply:
Renamed to `GenerateUpsertVisibilityTasks`